### PR TITLE
1693 - add github action to run jest tests

### LINKED ISSUE #1693: Testing - Set up github actions to run jest and phpunit tests
A/C
* Add Individual workflow  file for Jest ,  Php unit tests and linting
* Jest & Php tests status checks should be successful for PR to be successful
* Linting failures should not fail the PR

### PR DESCRIPTION
**Pull request by brindasasi**
resolves #1693 
A/C 

* Add individual workflow for jest , lint and php unit tests 
* make linting errors not to block merging
* jest & php unit test failures should block merge.

### CODE REVIEW

File: .github/workflows/node.js.yml
Comment (patcon) [inline .github/workflows/node.js.yml:7]: Since there's no real harm in enabling for all PRs, can we do that? There are some edge cases whether a PR against non-main branch is helpful. Thanks!

```suggestion
```

Ps, sorry, I know you're not asking for review yet, but worried I'll miss the review window since I'm not yet checking notifications every day :)

File: .github/workflows/node.js.yml
Comment (patcon) [inline .github/workflows/node.js.yml:41]: This was run in `frontend/` already, so no need to re-run. I think this actually runs an install as if not a monorepo, regenerating `frontend/common/package-lock.json` in the process, which shouldn't exist. So it's likely bad to run again

```suggestion
```

File: .github/workflows/node.js.yml
Comment (patcon) [inline .github/workflows/node.js.yml:49]:
```suggestion
```

File: .github/workflows/node.js.yml
Comment (patcon) [inline .github/workflows/node.js.yml:12]: Only needed because you're using it below.

```suggestion
    runs-on: ubuntu-latest
    env:
      NPM_VERSION: "8.3.0"
```

File: .github/workflows/node.js.yml
Comment (brindasasi) [reply .github/workflows/node.js.yml:7]: no worries. pls suggest as you them along . Thats the purpose of having draft PR -  to get some early feedback.

File: .github/workflows/lint.yml
Comment (tristan-orourke) [inline .github/workflows/lint.yml:36]: I think this might no longer be necessary? @patcon ?

File: .github/workflows/node.js.yml
Comment (tristan-orourke) [inline .github/workflows/node.js.yml:1]: Maybe this workflow should be title jest instead of node

File: .github/workflows/php.yml
Comment (tristan-orourke) [inline .github/workflows/php.yml:1]: And this workflow title phpunit instead of php

File: .github/workflows/php.yml
Comment (tristan-orourke) [inline .github/workflows/php.yml:38]: Are there any front-end dependencies here?

**Comment by patcon** [comment]
Hey @brindasasi! Just wanted to confirm that I'll review this in the morning :)

File: .github/workflows/lint.yml
Comment (brindasasi) [reply .github/workflows/lint.yml:36]: ok let me try removing it

Comment (patcon) [reply .github/workflows/lint.yml:36]: That's my understanding! The root issue is fixed in https://github.com/hydrogen-design-system/hydrogen.css/issues/72 🚀

Comment (patcon) [reply .github/workflows/lint.yml:36]: can be removed in other places in this PR too!

File: .github/workflows/node.js.yml
Comment (patcon) [reply .github/workflows/node.js.yml:1]:
```suggestion
name: Jest Tests
```

Comment (brindasasi) [reply .github/workflows/node.js.yml:1]: sure

File: .github/workflows/lint.yml
Comment (patcon) [inline .github/workflows/lint.yml:15]:
```suggestion
      - uses: actions/checkout@v2.4.0
```

File: .github/workflows/lint.yml
Comment (patcon) [inline .github/workflows/lint.yml:17]:
```suggestion
        uses: actions/setup-node@v2.5.1
```

File: .github/workflows/lint.yml
Comment (patcon) [inline .github/workflows/lint.yml:19]: can use nvmrc file in frontend

```suggestion
          node-version-file: frontend/.nvmrc
```

https://github.com/actions/setup-node/blob/main/docs/advanced-usage.md#node-version-file

File: .github/workflows/lint.yml
Comment (patcon) [inline .github/workflows/lint.yml:50]: Maybe this is worse for failures, but recently learned that this is possible for workspaces:

https://docs.npmjs.com/cli/v7/using-npm/workspaces#ignoring-missing-scripts

```suggestion
      - name: "Run ESlint: all workspaces"
        working-directory: frontend
        continue-on-error: true
        run: |
          npm run codegen --workspaces --if-present
          npm run lint --workspaces --if-present
```

Also, `continue-on-error`: does that raise an warning at all, or just skip it? I always use `if: ${{ always() }}` on subsequent steps (after one I think might fail, but that I was still to run, like more tests). I think `continue-on-error` will totally hide errors in PRs, instead of just moving past them in subsequent steps?

https://docs.github.com/en/actions/using-workflows/workflow-syntax-for-github-actions#jobsjob_idstepscontinue-on-error

> Prevents a job from failing when a step fails. Set to true to allow a job to pass when this step fails.

File: .github/workflows/node.js.yml
Comment (patcon) [inline .github/workflows/node.js.yml:15]:
```suggestion
      - uses: actions/checkout@v2.4.0
```

File: .github/workflows/node.js.yml
Comment (patcon) [inline .github/workflows/node.js.yml:17]:
```suggestion
        uses: actions/setup-node@v2.5.1
```

File: .github/workflows/lint.yml
Comment (patcon) [reply .github/workflows/lint.yml:50]: Sorry, this is confusing. Happy to stick with how we were doing it, but just remove `continue-on-error` like so:

```suggestion
      - name: "Run ESlint: common"
        working-directory: frontend/common
        if: ${{ always() }}
        run: |
          npm run codegen
          npm run lint

      - name: "Run ESlint: admin"
        working-directory: frontend/admin
        if: ${{ always() }}
        run: |
          npm run codegen
          npm run lint

      - name: "Run ESlint: talentSearch"
        working-directory: frontend/talentsearch
        if: ${{ always() }}
        run: |
          npm run codegen
          npm run lint
```

File: .github/workflows/phpunit.yml
Comment (patcon) [inline .github/workflows/phpunit.yml:23]:
```suggestion
      - uses: actions/checkout@2.4.0
```

File: .github/workflows/jest.yml
Comment (patcon) [inline .github/workflows/jest.yml:15]:
```suggestion
      - uses: actions/checkout@2.4.0
```

File: .github/workflows/jest.yml
Comment (patcon) [inline .github/workflows/jest.yml:17]:
```suggestion
        uses: actions/setup-node@v2.5.1
```

File: .github/workflows/node.js.yml
Comment (patcon) [inline .github/workflows/node.js.yml:35]:
```suggestion
          # No need to build since we only use admin's hydrogen.css.
          npm run h2-build
```

File: .github/workflows/lint.yml
Comment (brindasasi) [reply .github/workflows/lint.yml:50]: checkout this link :
https://github.com/GCTC-NTGC/gc-digital-talent/runs/5190508335?check_suite_focus=true 
Continue on error doesn't hide any errors. It just move on to next step without exiting the yml file and make the final status as "Success"

File: .github/workflows/node.js.yml
Comment (patcon) [inline .github/workflows/node.js.yml]: ditto

```suggestion
          npm run h2-build
```

File: .github/workflows/lint.yml
Comment (patcon) [reply .github/workflows/lint.yml:50]: sorry, when i say "hide errors" i don't mean that it hides them from logs when someone goes to look, but rather than it hides them in the status checks.

Screen Shot 2022-02-15 at 11 44 52 AM

So if I'm understanding your link correctly, it shows "run linters" as green checkmark, even when some steps failed. And so in the PR, it will show "run linters" as passing, and no one would know to check the logs.

Am I understanding that correctly?

Comment (patcon) [reply .github/workflows/lint.yml:50]: But maybe you could help me understand: would we want final status to be success when there are errors?

File: api/tests/UserTest.php
Comment (patcon) [inline api/tests/UserTest.php:14]: Maybe we could leave comments about why these are skipped? I am curious myself, but as it stands need to find someone to ask :)

**Comment by patcon** [comment]
Sorry, I think some of my comments were lost during the file rename. I'll re-add them, but wanted to clarify in case you were getting deja-vu 😄

File: .github/workflows/jest.yml
Comment (patcon) [inline .github/workflows/jest.yml:5]: Did you intend to remove `pull_request` triggers?

File: .github/workflows/lint.yml
Comment (patcon) [inline .github/workflows/lint.yml:5]:
```suggestion
  push:
    branches: [ main ]
  pull_request:
```

File: .github/workflows/phpunit.yml
Comment (patcon) [inline .github/workflows/phpunit.yml:3]:
```suggestion
on:
  push:
    branches: [ main ]
  pull_request:
```

File: .github/workflows/jest.yml
Comment (patcon) [inline .github/workflows/jest.yml:23]:
```suggestion
      - name: Upgrade NPM
```

File: .github/workflows/jest.yml
Comment (patcon) [inline .github/workflows/jest.yml:54]: I don't know if CSS if necessary for Jest, but if it is, then `npm run h2-bulld` needs to run for all subdirs

```suggestion
          npm run h2-build
          npm run codegen
          npm run test

      - name: "Run Unit Tests: admin"
        working-directory: frontend/admin
        run: |
          npm run h2-build
          npm run codegen
          npm run intl-compile
          npm run test

      - name: "Run Unit Tests: talentsearch"
        working-directory: frontend/talentsearch
        run: |
          npm run h2-build
          npm run codegen
          npm run intl-compile
          npm run test
```

File: .github/workflows/jest.yml
Comment (patcon) [inline .github/workflows/jest.yml:35]: I don't know if CSS if necessary for Jest, but if it is, then npm run h2-bulld needs to run for all subdirs

```suggestion
          npm run h2-build
```

File: .github/workflows/jest.yml
Comment (patcon) [inline .github/workflows/jest.yml:51]: ditto

```suggestion
          npm run h2-build
```

File: .github/workflows/jest.yml
Comment (patcon) [reply .github/workflows/jest.yml:36]:
```suggestion
```

File: .github/workflows/jest.yml
Comment (patcon) [reply .github/workflows/jest.yml:52]:
```suggestion
```

File: api/tests/UserTest.php
Comment (patcon) [reply api/tests/UserTest.php:14]:
```suggestion
        $this->markTestSkipped('Todo: Fix. https://github.com/GCTC-NTGC/gc-digital-talent/issues/1480');
```

File: api/tests/UserTest.php
Comment (patcon) [inline api/tests/UserTest.php:88]:
```suggestion
        $this->markTestSkipped('Todo: Fix. https://github.com/GCTC-NTGC/gc-digital-talent/issues/1480');
```

File: .github/workflows/phpunit.yml
Comment (patcon) [inline .github/workflows/phpunit.yml:26]:
```suggestion
      - uses: actions/checkout@v2.4.0
```

File: .github/workflows/jest.yml
Comment (patcon) [inline .github/workflows/jest.yml:16]:
```suggestion
      - uses: actions/checkout@v2.4.0
```

File: .github/workflows/lint.yml
Comment (patcon) [reply .github/workflows/lint.yml:51]:
```suggestion
      - name: "Setup: all workspaces"
        working-directory: frontend
        run: npm run codegen -- workspaces --if-present

      - name: "Run ESlint: all workspaces"
        working-directory: frontend
        continue-on-error: true
        run: npm run lint
```

File: .github/workflows/jest.yml
Comment (patcon) [inline .github/workflows/jest.yml:50]:
```suggestion
      - name: "Setup: all workspaces"
        working-directory: frontend
        run: |
          npm run codegen --workspaces --if-present
          npm run intl-compile --workspaces --if-present

      - name: "Run Unit Tests: all workspaces"
        working-directory: frontend
        run: npm run test --workspaces --if-present
```

File: .github/workflows/phpunit.yml
Comment (patcon) [inline .github/workflows/phpunit.yml:29]:
```suggestion
        uses: shivammathur/setup-php@2.17.0
```

File: .github/workflows/jest.yml
Comment (patcon) [inline .github/workflows/jest.yml:6]:
```suggestion
  push:
    branches: [ main ]
    paths:
      - .github/workflows/*jest*.yml
      - frontend/**
  pull_request:
    paths:
      - .github/workflows/*jest*.yml
      - frontend/**
```

File: .github/workflows/lint.yml
Comment (patcon) [inline .github/workflows/lint.yml:6]:
```suggestion
  push:
    branches: [ main ]
    paths:
      - .github/workflows/*lint*.yml
      - frontend/**
  pull_request:
    paths:
      - .github/workflows/*lint*.yml
      - frontend/**
```

**Review by patcon — APPROVED** [review]
LGTM! 🌮

File: .github/workflows/phpunit.yml
Comment (patcon) [inline .github/workflows/phpunit.yml:42]:
```suggestion
        run: vendor/bin/phpunit --verbose
```

**Comment by petertgiles** [comment]
> * jest & php unit test failures should block merge.

Are we sure we want to block merge on unit test failure?  In 99% of situations it would be a bad idea but I'd like for a human to be able to override the robot.

Otherwise, this looks great!  I can't wait for this to go live.

**Comment by brindasasi** [comment]
> > * jest & php unit test failures should block merge.
> 
> Are we sure we want to block merge on unit test failure? In 99% of situations it would be a bad idea but I'd like for a human to be able to override the robot.
> 
> Otherwise, this looks great! I can't wait for this to go live.

I believe it should block. In unforeseen situations you may still skip the tests like how the php user tests have been skipped currently. While skipping it , we would add the ticket number which we plan to fix it in future. 

Screen Shot 2022-02-15 at 5 13 57 PM

**Comment by patcon** [comment]
I can see merit to both blocking and not blocking, so either is good! Seems nice to start with the more restrictive, while being very eager to dial back if it's annoying :)

**Comment by patcon** [comment]
As per where @brindasasi and I left things, just waiting on final review from @tristan-orourke for merge :)

**Comment by tristan-orourke** [comment]
I'm happy to leave this review to @patcon, go ahead and merge!

**Pull request merged**
